crates/shared/src/account_balances/simulation.rs

```diff
@@ -80,9 +88,172 @@ impl Balances {
         tracing::trace!(?query, ?amount, ?simulation, "simulated balances");
         Ok(simulation)
     }
+}
+
+struct ZkSyncBalancesSimulator {
+    balances: contracts::support::Balances,
+    settlement: contracts::GPv2Settlement,
+    state_overrides: StateOverrides,
+    vault_relayer: H160,
+    vault: H160,
+    web3: Web3,
+}
+
+impl ZkSyncBalancesSimulator {
+    pub fn new(web3: Web3, settlement: H160, vault_relayer: H160, vault: H160) -> Self {
+        // pick a random address for the helper contract to avoid accidental collisions
+        let balances_address = H160::random();
+
+        // prepare state overrides that put the contract code of the helper contract at
+        // chosen address
+        let balances = contracts::support::Balances::raw_contract()
+            .deployed_bytecode
+            .to_bytes()
+            .unwrap();
+        let state_overrides: StateOverrides = [(
+            balances_address,
+            StateOverride {
+                code: Some(balances),
+                ..Default::default()
+            },
+        )]
+        .into_iter()
+        .collect();
+
+        Self {
+            balances: contracts::support::Balances::at(&web3, balances_address),
+            settlement: contracts::GPv2Settlement::at(&web3, settlement),
+            web3,
+            vault_relayer,
+            vault,
+            state_overrides,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl BalancesSimulator for ZkSyncBalancesSimulator {
+    #[instrument(skip_all)]
+    async fn simulate(&self, query: &Query, amount: Option<U256>) -> Result<Simulation> {
+        // prepare the call that does the balance computation
+        let balances_calldata = self
+            .balances
+            .balance(
+                (self.settlement.address(), self.vault_relayer, self.vault),
+                query.owner,
+                query.token,
+                amount.unwrap_or_default(),
+                Bytes(query.source.as_bytes()),
+                query
+                    .interactions
+                    .iter()
+                    .map(|i| (i.target, i.value, Bytes(i.call_data.clone())))
+                    .collect(),
+            )
+            .tx
+            .data
+            .context("no calldata for balance computation call")?;
+
+        // execute that call via the storage accessible interface to execute
+        // it in the context of the settlement contract
+        let storage_accessible_calldata = self
+            .settlement
+            .simulate_delegatecall(
```

Comment:
If we can call `simulate_delegatecall` on settlement directly, can we do the same in the other callsite (rather than instantiating a `StorageAccessible` instance there)? This should simplify the code in the other callsite.